add chat

RedstoneParkour opened the pull request:
also move player input to `_unhandled_input()` to prevent player movement / menu activation / etc. while chat is in focus

RedstoneParkour committed:
add chat

also move player input to _unhandled_input() to prevent player movement / menu activation / etc. while chat is in focus

SakuruWasTaken commented on Assets/Scenes/Columbia/main.gd:
```diff
@@ -21,8 +22,11 @@ enum server_packets {
+	CHAT = 15,
 }
 
+signal chat_message(message)
```
what is this line for?

SakuruWasTaken replied:
[> what is this line for?](https://github.com/OpenLWR/OpenLWR/pull/11#discussion_r1632410316)

everything else is looking good, but you still haven't answered my question

RedstoneParkour replied:
sorry for the late reply
that line defines the `chat_message` signal which is emitted at [line 1791](https://github.com/kcinnaJlol/OpenLWR/blob/3dcd20a4b503674b6c3bc1be48c4171c16775186/Assets/Scenes/Columbia/main.gd#L1791) and is connected to `_on_chat_message` by [Chat.gd, line 6](https://github.com/kcinnaJlol/OpenLWR/blob/3dcd20a4b503674b6c3bc1be48c4171c16775186/Assets/Scenes/Player/Chat.gd#L6)

SakuruWasTaken replied:
i see

SakuruWasTaken commented on Assets/Scenes/Columbia/main.gd:
```diff
@@ -9,6 +9,7 @@ enum client_packets {
 	ROD_SELECT_UPDATE = 11,
 	USER_LOGIN = 12,
 	SYNCHRONIZE = 14,
+	CHAT = 15,
```
it won't actually cause any problems, but we'd like to avoid having a client and a server packet with the same ID, just so that if you say "packet x" you don't also need to mention if it's a client packet or server packet

watchful5406 reviewed (changes requested):
chat should close when the user clicks anywhere else but it,
new chat messages should disappear after some time to avoid clogging up the screen (or make it possible to close

RedstoneParkour committed:
improve chat

add chat fading, max amount of chat messages (currently 200) and scroll to a new chat message when one arrives

RedstoneParkour committed:
change server-to-client chat packet id from 15 to 16

RedstoneParkour committed:
unfocus chat input when clicking outside it

RedstoneParkour committed:
change the auto-scroll hack to be more accurate

RedstoneParkour committed:
revert change; it doesn't work
